### ISSUE TITLE
All players showing as free agents

### ISSUE THREAD

**Issue by mdryden** [110yards/fantasy#192]
**Describe the bug**
All players on rosters and matchup pages are showing as if they are a free agent.  The player list may be ok for your league, if someone in your league has accessed it this week, but unless the data is fixed by morning, the caches will all be invalidated at 8am central and they will be affected too.

I can only assume this is a bug or change in the structure of the CFL's source data, as I can think of no way that the site would glitch out and set everyone to free agent.

I will investigate tomorrow if the issue is still ongoing.

**League ID**
n/a

**Roster name**
n/a

**Screenshots**
![Screenshot_20220619-221848.png](https://user-images.githubusercontent.com/5611490/174519356-df6e7228-c576-4537-9237-93a11422713c.png)

**Device information**
Affects mobile and desktop versions of the site.

**Comment by mdryden** [comment]
I don't know what happened here - by the time I got to it, the data was coming back clean from the CFL, but the data in my system was still tainted.

I suspect based on the logs that this happened last night around midnight.  I added an option to force update all players and everyone is back to working now.

**Comment by mdryden** [comment]
Still happening; a few players have gone back to free agent status.

**Comment by mdryden** [comment]
Digging through the code, I'd guess that the CFL's API is dropping players from the list - when that happens, the system will assume the player is no longer on a team (that's how it actually happens when a player is released), and then set the player as a free agent.  Looking into what I can do about these false free agents.

**Comment by mdryden** [comment]
Found the issue here.

First, the CFL API has a weird quirk where if I ask for players and teams to be returned, it only returns players who are currently on a team.  This has been a problem in the past, because I never know when players become free agents.  I resolved that issue this season by comparing the list of players from the CFL with the current list in the site - if a player exists in my data, but not the CFL's, they are set to a free agent.

Second, there was a bug in my code when I set them to be a free agent - after changing the team, I didn't recalculate the hash (which is used for detecting changes) - this meant that if they player appeared again on the same team, the site didn't see any change, and left them as a free agent.

What I think happened is that the CFL's API bugged out and dropped some players from the responses, causing the system to think they had been released.  Once they had been "released" in the system, they were never set back, even if they were in the CFL's data during the next update.

I've fixed that bug; if a player is released and renewed by the same team, (or falsely released and reappears), the site will now properly restore their team information.

Second, I've added an extra layer to the update to try and avoid accidentally marking players as free agents, if they temporarily disappear from the player list - the site will now fetch that player individually from the CFL and check their team that way.  I suspect that the issue is only on the bulk fetch side, so I think this will probably prevent it from happening again.
